File: c/tutorial/go/tutorial.go
// FlintDB Go Tutorial
package main

import (
	"fmt"
	"os"

	flintdb "flintdb-tutorial/flintdb"
)

// tutorialTableCreate demonstrates how to create a table, define a schema, and insert data
func tutorialTableCreate() error {
	fmt.Println("--- Running tutorialTableCreate ---")

	tablename := "./temp/tutorial_customer.flintdb"
	flintdb.TableDrop(tablename)

	// 1. Define the table schema
	meta, err := flintdb.NewMeta(tablename)
	if err != nil {
		return err
	}
	defer meta.Close()

	if err := meta.AddColumn("id", flintdb.VARIANT_INT64, 0, 0, flintdb.SPEC_NOT_NULL, "0", "PRIMARY KEY"); err != nil {
		return err
	}
	if err := meta.AddColumn("name", flintdb.VARIANT_STRING, 50, 0, flintdb.SPEC_NOT_NULL, "", "Customer name"); err != nil {
		return err
	}
	if err := meta.AddColumn("age", flintdb.VARIANT_INT32, 0, 0, flintdb.SPEC_NOT_NULL, "0", "Customer age"); err != nil {
		return err
	}

	if err := meta.AddIndex(flintdb.PRIMARY_NAME, []string{"id"}); err != nil {
		return err
	}
	if err := meta.AddIndex("ix_age", []string{"age"}); err != nil {
		return err
	}

	sql, err := meta.ToSQL()
	if err != nil {
		return err
	}
	fmt.Printf("Table schema SQL:\n%s\n\n", sql)

	// 2. Open the table
	table, err := flintdb.TableOpen(tablename, flintdb.FLINTDB_RDWR, meta)
	if err != nil {
		return err
	}
	defer table.Close()

	// 3. Insert data rows
	fmt.Println("Inserting 3 rows...")
	for i := 0; i < 3; i++ {
		row, err := table.CreateRow()
		if err != nil {
			return err
		}

		if err := row.SetInt64ByName("id", int64(i+1)); err != nil {
			row.Free()
			return err
		}

		name := fmt.Sprintf("Customer %d", i+1)
		if err := row.SetStringByName("name", name); err != nil {
			row.Free()
			return err
		}

		if err := row.SetInt32ByName("age", int32(30+i)); err != nil {
			row.Free()
			return err
		}

		rowid, err := table.Insert(row)
		row.Free()
		if err != nil {
			return err
		}
		_ = rowid
	}

	fmt.Println("Successfully created table and inserted data.\n")
	return nil
}

// tutorialTableFind demonstrates how to find and read data from an existing table
func tutorialTableFind() error {
	fmt.Println("--- Running tutorialTableFind ---")

	tablename := "./temp/tutorial_customer.flintdb"

	// 1. Open the table in read-only mode
	table, err := flintdb.TableOpen(tablename, flintdb.FLINTDB_RDONLY, nil)
	if err != nil {
		return err
	}
	defer table.Close()

	// 2. Find data using a WHERE clause
	fmt.Println("Finding rows where age >= 31:")
	cursor, err := table.Find("WHERE age >= 31")
	if err != nil {
		return err
	}
	defer cursor.Close()

	// 3. Iterate through the cursor
	for {
		rowid, err := cursor.Next()
		if err != nil {
			return err
		}
		if rowid < 0 {
			break
		}

		row, err := table.Read(rowid)
		if err != nil {
			return err
		}

		row.Print()
	}

	fmt.Println("\nSuccessfully found and read data.\n")
	return nil
}

// tutorialTsvCreate demonstrates how to create a TSV file and write data to it
func tutorialTsvCreate() error {
	fmt.Println("--- Running tutorialTsvCreate ---")

	filepath := "./temp/tutorial_products.tsv"
	flintdb.GenericFileDrop(filepath)

	// 1. Define the schema for the TSV file
	meta, err := flintdb.NewMeta(filepath)
	if err != nil {
		return err
	}
	defer meta.Close()

	// Configure for TSV
	meta.SetFormatTSV()

	if err := meta.AddColumn("product_id", flintdb.VARIANT_INT32, 0, 0, flintdb.SPEC_NOT_NULL, "", ""); err != nil {
		return err
	}
	if err := meta.AddColumn("product_name", flintdb.VARIANT_STRING, 100, 0, flintdb.SPEC_NOT_NULL, "", ""); err != nil {
		return err
	}
	if err := meta.AddColumn("price", flintdb.VARIANT_DOUBLE, 0, 0, flintdb.SPEC_NOT_NULL, "", ""); err != nil {
		return err
	}

	// 2. Open the generic file
	file, err := flintdb.GenericFileOpen(filepath, flintdb.FLINTDB_RDWR, meta)
	if err != nil {
		return err
	}
	defer file.Close()

	// 3. Write data rows
	fmt.Println("Writing 3 rows to TSV...")
	for i := 0; i < 3; i++ {
		row, err := file.CreateRow()
		if err != nil {
			return err
		}

		if err := row.SetInt32ByName("product_id", int32(101+i)); err != nil {
			row.Free()
			return err
		}

		name := fmt.Sprintf("Product-%c", 'A'+i)
		if err := row.SetStringByName("product_name", name); err != nil {
			row.Free()
			return err
		}

		if err := row.SetDoubleByName("price", 9.99*float64(i+1)); err != nil {
			row.Free()
			return err
		}

		if err := file.Write(row); err != nil {
			row.Free()
			return err
		}
		row.Free()
	}

	fmt.Println("Successfully created TSV file.\n")
	return nil
}

// tutorialTsvFind demonstrates how to read data from a TSV file
func tutorialTsvFind() error {
	fmt.Println("--- Running tutorialTsvFind ---")

	filepath := "./temp/tutorial_products.tsv"

	// 1. Open the TSV file in read-only mode
	file, err := flintdb.GenericFileOpen(filepath, flintdb.FLINTDB_RDONLY, nil)
	if err != nil {
		return err
	}
	defer file.Close()

	// 2. Find rows
	fmt.Println("Reading rows from TSV where product_id >= 102:")
	cursor, err := file.Find("WHERE product_id >= 102")
	if err != nil {
		return err
	}
	defer cursor.Close()

	// 3. Iterate through the cursor
	for {
		row, err := cursor.Next()
		if err != nil {
			return err
		}
		if row == nil {
			break
		}

		row.Print()
		row.Free()
	}

	fmt.Println("\nSuccessfully read from TSV file.\n")
	return nil
}

// tutorialTableUpdateDelete demonstrates how to update and delete rows in a table
func tutorialTableUpdateDelete() error {
	fmt.Println("--- Running tutorialTableUpdateDelete ---")

	tablename := "./temp/tutorial_customer.flintdb"

	// 1. Open the table in read-write mode
	table, err := flintdb.TableOpen(tablename, flintdb.FLINTDB_RDWR, nil)
	if err != nil {
		return err
	}
	defer table.Close()

	// 2. Find a row to update
	fmt.Println("Finding and updating Customer with age = 30:")
	cursor, err := table.Find("WHERE age = 30")
	if err != nil {
		return err
	}
	defer cursor.Close()

	rowid, err := cursor.Next()
	if err != nil {
		return err
	}

	if rowid > -1 {
		oldRow, err := table.Read(rowid)
		if err != nil {
			return err
		}

		fmt.Println("Before update:")
		oldRow.Print()

		// Create updated row
		newRow, err := table.CreateRow()
		if err != nil {
			return err
		}
		defer newRow.Free()

		// Copy and modify
		if err := newRow.SetInt64ByName("id", 1); err != nil {
			return err
		}
		if err := newRow.SetStringByName("name", "Updated Customer"); err != nil {
			return err
		}
		if err := newRow.SetInt32ByName("age", 35); err != nil {
			return err
		}

		// Update at specific rowid
		if err := table.UpdateAt(rowid, newRow); err != nil {
			return err
		}

		fmt.Println("After update:")
		updatedRow, err := table.Read(rowid)
		if err != nil {
			return err
		}
		updatedRow.Print()
	}

	// 3. Delete a row
	fmt.Println("\nDeleting Customer with id = 3:")
	cursor2, err := table.Find("WHERE id = 3")
	if err != nil {
		return err
	}
	defer cursor2.Close()

	rowid, err = cursor2.Next()
	if err != nil {
		return err
	}

	if rowid > -1 {
		if err := table.DeleteAt(rowid); err != nil {
			return err
		}
		fmt.Printf("Successfully deleted row at rowid %d\n", rowid)
	}

	// 4. Verify deletion
	fmt.Println("\nRemaining customers:")
	cursor3, err := table.Find("")
	if err != nil {
		return err
	}
	defer cursor3.Close()

	for {
		rowid, err := cursor3.Next()
		if err != nil {
			return err
		}
		if rowid < 0 {
			break
		}

		row, err := table.Read(rowid)
		if err != nil {
			return err
		}
		row.Print()
	}

	fmt.Println("\nSuccessfully updated and deleted rows.\n")
	return nil
}

// tutorialFilesort demonstrates how to use filesort for external sorting
// Note: Full filesort bindings require more complex CGo setup
// This is a placeholder for the feature
func tutorialFilesort() error {
	fmt.Println("--- Filesort feature available in C API ---")
	fmt.Println("(Go bindings for filesort require additional CGo setup)\n")
	return nil
}

func main() {
	// Create temp directory if it doesn't exist
	if _, err := os.Stat("./temp"); os.IsNotExist(err) {
		os.Mkdir("./temp", 0755)
	}

	if err := tutorialTableCreate(); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}

	if err := tutorialTableFind(); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}

	if err := tutorialTableUpdateDelete(); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}

	if err := tutorialTsvCreate(); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}

	if err := tutorialTsvFind(); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}

	if err := tutorialFilesort(); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}

	fmt.Println("All tutorial steps completed successfully.")
}

File: c/tutorial/zig/tutorial.zig
/// FlintDB Zig Tutorial
const std = @import("std");
const flintdb = @import("flintdb.zig");

// const c = @cImport({
//     @cInclude("flintdb.h");
// });

fn tutorialTableCreate() !void {
    std.debug.print("--- Creating Customer Table ---\n", .{});

    const tablename = "./temp/tutorial_customer.flintdb";
    flintdb.Table.drop(tablename);

    // Create metadata
    var meta = try flintdb.Meta.init(tablename);
    defer meta.deinit();

    // Add columns
    try meta.addColumn("id", flintdb.VariantType.INT64, 0, 0, 1, "0", "PRIMARY KEY");
    try meta.addColumn("name", flintdb.VariantType.STRING, 50, 0, 1, "", "Customer name");
    try meta.addColumn("age", flintdb.VariantType.INT32, 0, 0, 1, "0", "Customer age");

    // Add indexes
    try meta.addIndex(flintdb.PRIMARY_KEY, &[_][]const u8{"id"});
    try meta.addIndex("ix_age", &[_][]const u8{"age"});

    // Print schema
    var sql_buf: [2048]u8 = undefined;
    const sql = try meta.toSQL(&sql_buf);
    std.debug.print("Schema SQL:\n{s}\n\n", .{sql});

    // Open table and insert data
    var table = try flintdb.Table.open(tablename, .FLINTDB_RDWR, &meta);
    defer table.close();

    std.debug.print("Inserting 3 customers (using column names)...\n", .{});
    var i: usize = 0;
    while (i < 3) : (i += 1) {
        var row = try table.createRow();
        defer row.free();

        // Using column names instead of indices!
        try row.setInt64ByName("id", @intCast(i + 1));

        var name_buf: [50]u8 = undefined;
        const name = try std.fmt.bufPrint(&name_buf, "Customer {d}", .{i + 1});
        try row.setStringByName("name", name);

        try row.setInt32ByName("age", @intCast(30 + i));

        const rowid = try table.insert(row);
        std.debug.print("  Inserted rowid: {d}\n", .{rowid});
    }
    std.debug.print("✓ Table created successfully\n\n", .{});
}

fn tutorialTableFind() !void {
    std.debug.print("--- Finding Customers (age >= 31) ---\n", .{});

    const tablename = "./temp/tutorial_customer.flintdb";
    var table = try flintdb.Table.open(tablename, .FLINTDB_RDONLY, null);
    defer table.close();

    var cursor = try table.find("WHERE age >= 31");
    defer cursor.close();

    while (try cursor.next()) |rowid| {
        const row = try table.read(rowid);
        row.print();
    }

    std.debug.print("✓ Query completed\n\n", .{});
}

fn tutorialTsvCreate() !void {
    std.debug.print("--- Creating TSV File ---\n", .{});

    const filepath = "./temp/tutorial_products.tsv";
    flintdb.GenericFile.drop(filepath);

    var meta = try flintdb.Meta.init(filepath);
    defer meta.deinit();

    // Configure for TSV
    const tsv = "tsv";
    @memcpy(meta.inner.format[0..tsv.len], tsv);
    meta.inner.format[tsv.len] = 0;
    meta.inner.delimiter = '\t';
    meta.inner.absent_header = 1;

    // Add columns
    try meta.addColumn("product_id", flintdb.VariantType.INT32, 0, 0, 1, "", "");
    try meta.addColumn("product_name", flintdb.VariantType.STRING, 100, 0, 1, "", "");
    try meta.addColumn("price", flintdb.VariantType.DOUBLE, 0, 0, 1, "", "");

    var file = try flintdb.GenericFile.open(filepath, .FLINTDB_RDWR, &meta);
    defer file.close();

    std.debug.print("Writing 3 products (using column names)...\n", .{});
    var i: usize = 0;
    while (i < 3) : (i += 1) {
        var row = try file.createRow();
        defer row.free();

        // Use column names like C example
        try row.setInt32ByName("product_id", @intCast(101 + i));

        var name_buf: [100]u8 = undefined;
        const ch: u8 = @intCast('A' + i);
        const name = try std.fmt.bufPrint(&name_buf, "Product-{c}", .{ch});
        try row.setStringByName("product_name", name);

        const price = 9.99 * @as(f64, @floatFromInt(i + 1));
        try row.setDoubleByName("price", price);

        try file.write(row);
        std.debug.print("  Written: Product-{c}, ${d:.2}\n", .{ ch, price });
    }
    std.debug.print("✓ TSV file created\n\n", .{});
}

fn tutorialTableQuery() !void {
    std.debug.print("--- Advanced Query Examples ---\n", .{});

    const tablename = "./temp/tutorial_customer.flintdb";
    var table = try flintdb.Table.open(tablename, .FLINTDB_RDONLY, null);
    defer table.close();

    // Query: All customers (using column names!)
    std.debug.print("All customers (accessed by column names):\n", .{});
    var cursor1 = try table.find("");
    defer cursor1.close();
    while (try cursor1.next()) |rowid| {
        const row = try table.read(rowid);

        var name_buf: [50]u8 = undefined;
        // Access by column name instead of index!
        const id = try row.getInt64ByName("id");
        const name = try row.getStringByName("name", &name_buf);
        const age = try row.getInt32ByName("age");

        std.debug.print("  ID={d}, Name={s}, Age={d}\n", .{ id, name, age });
    }

    std.debug.print("\n✓ Queries completed\n\n", .{});
}

fn tutorialTableUpdateDelete() !void {
    std.debug.print("--- Table Update/Delete Operations ---\n", .{});

    const tablename = "./temp/tutorial_customer.flintdb";
    var table = try flintdb.Table.open(tablename, .FLINTDB_RDWR, null);
    defer table.close();

    std.debug.print("(Update/delete operations require additional binding implementation)\n", .{});
    std.debug.print("✓ Feature available in C API\n\n", .{});
}

fn tutorialFilesort() !void {
    std.debug.print("--- Filesort for External Sorting ---\n", .{});
    std.debug.print("(Filesort bindings require additional Zig implementation)\n", .{});
    std.debug.print("✓ Feature available in C API\n\n", .{});
}

pub fn main() !void {
    defer flintdb.cleanup();

    // Ensure temp directory exists
    var cwd = std.fs.cwd();
    cwd.makeDir("temp") catch |err| switch (err) {
        error.PathAlreadyExists => {},
        else => return err,
    };

    try tutorialTableCreate();
    try tutorialTableFind();
    try tutorialTableQuery();
    try tutorialTableUpdateDelete();
    try tutorialTsvCreate();
    try tutorialFilesort();

    std.debug.print("✨ All tutorials completed successfully!\n", .{});
}
